SSA/Projects/CIRCT/Comb/Comb.lean

```diff
@@ -0,0 +1,617 @@
+    {n : Nat} → HVector f (List.replicate n a) → List (f a)
+  | 0, _ => []
+  | n+1, HVector.cons x xs => x :: replicateToList xs
+
+def ofString? (s : String) : Option CombOp.IcmpPredicate :=
+  match s with
+  | "eq" => some .eq
+  | "ne" => some .ne
+  | "slt" => some .slt
+  | "sle" => some .sle
+  | "sgt" => some .sgt
+  | "sge" => some .sge
+  | "ult" => some .ult
+  | "ule" => some .ule
+  | "ugt" => some .ugt
+  | "uge" => some .uge
+  | _     => none
+
+def_denote for Comb where
+  | .add _ _ => fun xs => CombOp.add (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .and _ _ => fun xs => CombOp.and (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .concat _ => fun xs => CombOp.concat xs
+  | .divs _ => fun xs => CombOp.divs xs
+  | .divu _ => fun xs => CombOp.divu xs
+  | .extract _ n => fun xs => CombOp.extract xs n
+  | .icmp p _ => fun xs =>
+      let p' := Option.get! (ofString? p)
+      CombOp.icmp p' xs
+  | .mods _ => fun xs => CombOp.mods xs
+  | .modu _ => fun xs => CombOp.modu xs
+  | .mul _ _ => fun xs => CombOp.mul (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .mux _ => fun xs => CombOp.mux xs
+  | .or _ _ => fun xs => CombOp.or (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .parity _ => fun xs => CombOp.parity xs
+  | .replicate _ n => fun xs => CombOp.replicate xs n
+  | .shl _ => fun xs => CombOp.shl xs
+  | .shrs _ => fun xs => CombOp.shrs xs
+  | .shru _ => fun xs => CombOp.shru xs
+  | .sub _ => fun xs => CombOp.sub xs
+  | .xor _ _ => fun xs => CombOp.xor (HVector.replicateToList (f := TyDenote.toType) xs)
+
+end Dialect
+
+/-- we currently do not support the parsing of heterogeneous vectors -/
+def mkTy : MLIR.AST.MLIRType φ → MLIR.AST.ExceptM Comb Ty
+  | MLIR.AST.MLIRType.undefined s => do
+    match s.splitOn "_" with
+    | ["Bool"] =>
+      return .bool
+    | ["Nat"] =>
+      return .nat
+    | ["BitVec", r] =>
+      return .bv (r.toNat!)
+    | ["IcmpPred"] =>
+      -- match icmp
+      return .icmpPred
+    | _ => throw .unsupportedType
+  | _ => throw .unsupportedType
+
+instance instTransformTy : MLIR.AST.TransformTy Comb 0 where
+  mkTy := mkTy
+
+
+def add {Γ : Ctxt _} (arity: Nat) (a : HVector Γ.Var (List.replicate arity (.bv w))) : Expr (Comb) Γ .pure (.bv w) :=
+  Expr.mk
+    (op := .add w arity)
+    (ty_eq := by rfl)
+    (eff_le := by constructor)
+    (args := a) -- cast a to the right type
+    (regArgs := .nil)
+
+def and {Γ : Ctxt _} (arity: Nat) (a : HVector Γ.Var (List.replicate arity (.bv w))) : Expr (Comb) Γ .pure (.bv w) :=
+  Expr.mk
+    (op := .and w arity)
+    (ty_eq := by rfl)
+    (eff_le := by constructor)
+    (args := a)
+    (regArgs := .nil)
+
+def concat {l : List Nat} {Γ : Ctxt _} (ls : Γ.Var (.hList l)): Expr (Comb) Γ .pure (.bv (l.sum)) :=
+  Expr.mk
+    (op := .concat l)
+    (ty_eq := rfl)
+    (eff_le := by constructor)
+    (args := .cons ls <| .nil)
+    (regArgs := .nil)
```

Comment:
Here again you've got a huge number of boilerplate definitions. Do we actually need these? I know we have similar constructors in other dialects where we didn't bother writing a parser, but for Comb you did write a parser, so there should be no need for these defs.

Of course, I see you've used them in the definition of said parser, but you should be able to write that much more concisely.

Reply:
could you please make an example? I don't understand how these can be avoided

Reply:
Im guessing that the definitions can just be completely inlined, because they are only used in the parser and not needed again.

Reply:
Okay I'll refactor the parser to extract some functions based on the number of argumens and inline this as well